### ISSUE TITLE
How to use BatchNorm

### ISSUE THREAD

**Issue by MarkKremer** [gorgonia/gorgonia#408]
Hello :wave: 

I'm trying to port a neural net which makes use of some BatchNorm layers. But I'm confused how to use the BatchNorm op in Gorgonia.

### Which nodes are trainable?
I'm assuming the `scale` (`γ`) and `bias` (`β`). However, the comment above the `BatchNorm` function tells a different story:
```golang
// BatchNorm applies a batchnormalization. This operator can be used in forward pass or for training.
// In an evaluation only, the "op" output can be discared.
// In training phase, γ, β can be discarded and the op should be used.
func BatchNorm(x, scale, bias *Node, momentum, epsilon float64) (retVal, γ, β *Node, op *BatchNormOp, err error) {
```
How is `BatchNormOp` used during training? I see that the `SetTraining` and `SetTesting` methods exist, but it is set to training by default already.

### How do I save the op's state?
The `mean`, `variance` etc. aren't exported or accessible.

### The default scale and bias keep the network from learning
When I started using the op I set the `scale` and `bias` to `nil`. `BatchNorm` will create these tensors for me and return them as `γ` and `β`. Using these defaults, my network doesn't train at all. If I manually initialize the `scale` with ones and the `bias` with zeroes it does work. Is this just weird or am I doing something wrong?

### How to do validation during training?
Here's a simplified version of my training loop:
```golang
for i := 1; i <= epochs; i++ {
    m.SetTraining()
    m.Train(trainData.Load())
    cost, acc := m.Metrics()

    m.SetInference()
    m.Predict(validationData.Load())
    valCost, valAcc := m.Metrics()

    // print costs and accuracies
}
```

Every time I call the `SetTraining` function on the `BatchNorm` op it will implicitly call `Reset` too. With my loop, the train accuracy still goes up, but the validation accuracy stays around its initial value.

Thanks for the help :)

**Comment by matthiasak** [comment]
Hi @MarkKremer -- just curious if you ultimately found some guidance on this?

**Comment by bdleitner** [comment]
Saw today's comment and was reminded of this.  I also was looking for examples or guidance on using BatchNorm and the difference in setting it up for training / testing.  IIRC the setups in test were not illuminating, though I admit I haven't looked recently.

**Comment by MarkKremer** [comment]
It has been a long time :sweat_smile: . I think I found some bugs in the op (some of which I mentioned in this issue) but I'm not sure if I completely fixed it in the end and if those fixes were correct. See my old branches:
- https://github.com/MarkKremer/gorgonia/commits/batchnorm-2 ([diff](https://github.com/MarkKremer/gorgonia/compare/9f5ade00365405e5ee1ab5befd99d9f86eeb9bf8...1da3e8b0bbc6d2ac50c4d8f89bb629ff1486146f))
- https://github.com/MarkKremer/gorgonia/commits/batchnorm ([diff](https://github.com/MarkKremer/gorgonia/compare/ee392a700a0d4a7540c21996288d1ff4c0e0dc09...5ae8a408cc032282cfdcfdf32e094ceeb0e24629))

I think the `batchnorm` branch was my first attempt with a bunch of comments to figure out what things were doing. The `batchnorm-2` branch is a tidied up version of that (possibly incomplete) + some other changes with the intention of making a PR.

I also have a vague memory that one of the ops didn't match the CUDA version of that op. That could have been `BatchNorm`.

To answer my own questions as best I can:

**Which nodes are trainable?**
`γ` and `β`

**How is BatchNormOp used during training? / How to do validation during training?**
I think I was using `op.SetTraining()` and `op.SetTesting()` in the way you would expect (but see fixes in my branch). But in some other parts of my projects I wasn't. So I'm not sure anymore.

**How do I save the op's state?**
I don't know. `γ` and `β` can be saved. But not sure about `mean` and `variance`.

**The default scale and bias keep the network from learning**
I think this is still correct. See my changes to `scale` and `bias` in my `batchnorm-2` branch. You can pass `nil` as parameter and add the returned `γ` and `β` to the learnables.


Hope this helps.

**Comment by dcu** [comment]
hello, BN is currently broken for evaluation mode due to a problem when calculating the gradient. We're working to fix it this week.

Training works well but you'll need to use the master to be able to access the mean and variance: https://github.com/gorgonia/gorgonia/blob/master/op_nn.go#L1256
